src/main/java/org/wildfly/common/function/ThreadLocalStack.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2021 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.common.function;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+/**
+ * A thread local stack data structure. In order to avoid memory churn the underlying
+ * ArrayDeque is never freed. If we remove the deque when it is empty then this results
+ * in excessive deque allocations.
+ */
+public class ThreadLocalStack<E> {
+
+    private static final Object NULL_VALUE = new Object();
+
+    private final ThreadLocal<Deque<Object>> deque = new ThreadLocal<Deque<Object>>() {
+        @Override
+        protected ArrayDeque<Object> initialValue() {
+            return new ArrayDeque<>();
+        }
+    };
```

Comment:
This is relatively minor but you should be able to reduce LOC as well as the number of classes by doing this instead:

```suggestion
    private final ThreadLocal<Deque<Object>> deque = ThreadLocal.withInitial(ArrayDeque::new);
```